packages/cursorless-vscode/src/keyboard/grammar/getAcceptableTokenTypes.ts

```diff
@@ -43,39 +66,80 @@ function getMetadata<T extends keyof KeyboardCommandHandler>(
+
 /**
- * Given a state, returns all root states that are reachable from it. We use this
- * to find out which top-level rules want a given token type.
+ * Given a state, returns all root states that are reachable from it, including
+ * partially filled out arguments to the given rule. We use this to find out
+ * which top-level rules want a given token type.
  *
  * @param state The state to get the root states of
+ * @param lastSymbol The partially filled out symbol that is currently being
+ * constructed
+ * @param visitCounts A map of states to the number of times they've been
+ * visited. We use this to prevent infinite recursion
+ * @param roots The list of root states
  * @returns A list of root states
  */
-function getRootStates(state: nearley.State) {
-  /** A queue of states to process; ensures we don't try to process state twice */
-  const queue = new UniqueWorkQueue(state);
-  const roots: State[] = [];
+function computeRootStatePartialArgs(
+  state: nearley.State,
+  lastSymbol: any = NEXT,
+  visitCounts = new DefaultMap<State, number>(() => 0),
+  roots: { state: State; partialArg: any }[] = [],
+) {
+  const visitCount = visitCounts.get(state);
+  if (visitCount > MAX_VISITS) {
+    return roots;
+  }
+  visitCounts.set(state, visitCount + 1);
+
+  let partialArg: any;
+  try {
+    const argList = [...getCompletedSymbols(state), lastSymbol];
+    argList.push(
+      ...times(state.rule.symbols.length - argList.length, () => MISSING),
+    );
+    partialArg = state.rule.postprocess?.(argList) ?? argList;
+  } catch (err) {
+    partialArg = MISSING;
+  }
 
-  for (const state of queue) {
-    queue.push(...state.wantedBy);
+  if (state.wantedBy.length === 0) {
+    roots.push({ state, partialArg });
+  }
 
-    if (state.wantedBy.length === 0) {
-      roots.push(state);
-    }
+  for (const parent of state.wantedBy) {
+    // If we're not a root state, we recurse on all states that want this state,
+    // passing them our partial argument so that they can use it when constructing
+    // their partial argument
+    computeRootStatePartialArgs(parent, partialArg, visitCounts, roots);
   }
 
   return roots;
 }
 
 /**
- * Given a root state, returns a partial argument for the command that the state
- * represents. We could use this info to display which arguments have already
- * been filled out while a user is typing a command.
- * @param state A root state
- * @returns A partial argument for the command that the state represents
+ * Given a state, returns a list of the symbols that have already been
+ * completed.
+ *
+ * @param state A state
+ * @returns A list of completed symbols
  */
-function computePartialArg<T extends keyof KeyboardCommandHandler>(
-  _state: nearley.State,
-): Partial<Record<T, any>> {
-  // FIXME: Fill this out
-  return {};
+function getCompletedSymbols(state: nearley.State) {
+  let currentState = state;
+  const completedSymbols: any[] = [];
+
+  while (currentState.dot > 0) {
```

Comment:
naive question because my typescript is still weak: does mutating currentState here mutate state? (what is a value type and what is a reference type? sigh, sorry.)

Reply:
Sorry where do we mutate `currentState`?

Reply:
oh, i see i misread something. never mind, ignore me.

Reply:
Fwiw primitives are value types, everything else is reference. (keeping in mind that strings are considered primitive in that statement). Tho I think strings are immutable so doesn't really matter. But I'd need to double check that